Add support for WebAuthn PRF extension

emlun opened the pull request:
Transplanted from https://bugzilla.mozilla.org/show_bug.cgi?id=1863819 and https://phabricator.services.mozilla.com/D211528, and refined from rough prototype state.

jschanck commented on src/ctap2/attestation.rs:
```diff
@@ -1126,4 +1163,140 @@ pub mod test {
         }
     }
+
+    mod hmac_secret {
```
We run some CI workflows without a complete crypto backend, so we need to disable these tests with `#[cfg(not(feature = "crypto_dummy"))]`.

emlun replied:
Done! https://github.com/mozilla/authenticator-rs/pull/337/commits/5fb07b15dcfe34600ca9ec28eca3c05b5f7c5fba

jschanck commented on src/ctap2/attestation.rs:
```diff
@@ -23,6 +25,41 @@ pub enum HmacSecretResponse {
+                    .try_into()
+                    .map_err(|_| CryptoError::WrongSaltLength)?,
+                output2: Some(output2)
+                    .filter(|o2| !o2.is_empty())
```
nit: you can simplify this a bit using bool::then `(output2.len() > 0).then(|| output2.try_into().map_err(|_| CryptoError::WrongSaltLength)).transpose()?`

emlun replied:
Thanks, done! https://github.com/mozilla/authenticator-rs/pull/337/commits/3c04ddabb23310fde06e704175822500a69d5eae

jschanck commented on src/ctap2/commands/get_assertion.rs:
```diff
@@ -67,17 +68,100 @@ impl UserVerification for GetAssertionOptions {
+    HmacGetSecret(HmacSecretExtension),
+    /// hmac-secret input is to be calculated from PRF inputs, but we haven't yet identified which eval or evalByCredential entry to use.
+    PrfUninitialized(AuthenticationExtensionsPRFInputs),
+    /// prf client input with no eval or matchin evalByCredential entry.
```
"matchin" -> "matching"

emlun replied:
Thanks, done! https://github.com/mozilla/authenticator-rs/pull/337/commits/0f440aac616f12526801274a2de76e7518d9acb1

jschanck commented on src/ctap2/mod.rs:
```diff
@@ -672,6 +660,35 @@ pub fn sign<Dev: FidoDevice>(
             return false;
         }
 
+        // Third, use the shared secret in the extensions, if requested
```
I'm not sure what sequence "Third" referred to in the original, let's remove it.

emlun replied:
Done! https://github.com/mozilla/authenticator-rs/pull/337/commits/0f440aac616f12526801274a2de76e7518d9acb1

jschanck commented on src/ctap2/commands/get_assertion.rs:
```diff
@@ -121,10 +218,14 @@ impl Serialize for HmacSecretExtension {
         if let Some(calc) = &self.calculated_hmac {
-            let mut map = serializer.serialize_map(Some(3))?;
+            let mut map =
+                serializer.serialize_map(Some(3 + self.pin_protocol.map(|_| 1).unwrap_or(0)))?;
```
nit: `serializer.serialize_map(Some(self.pin_protocol.map_or(3, |_| 4)))?` or maybe         `serializer.serialize_map(Some(if self.pin_protocol.is_some() { 4 } else { 3 }) )?;`

emlun replied:
I started to address this, but then quickly found myself thinking it would be much better to not have this problem in the first place: https://github.com/mozilla/authenticator-rs/pull/338 :smile:

jschanck commented on src/ctap2/server.rs:
```diff
@@ -364,19 +367,144 @@ pub struct AuthenticationExtensionsClientInputs {
+
+    /// Convert a PRF eval input to an hmac-secret salt input.
+    fn eval_to_salt(eval: &[u8]) -> [u8; 32] {
+        sha256(
```
nit: I'm not sure the `sha256` function is helping us here. This could be:
```
        let mut hasher = Sha256::new();
        hasher.update(b"WebAuthn PRF\x00");
        hasher.update(eval);
        hasher.finalize().into()
```

emlun replied:
There is one more use of the function in this file, and we could use it in [`Pin::for_pin_token`](https://github.com/mozilla/authenticator-rs/blob/b21f8fa7c5268f271bd598a05885838a20a8c12d/src/ctap2/commands/client_pin.rs#L636) as well. How about moving the `sha256` function to the `util` module instead?

emlun replied:
Oh, I realize now that I completely missed the point you meant. :slightly_smiling_face: Agreed, I'll change this to use `Sha256::new_with_prefix(...).chain(...)` instead.

emlun replied:
Done: https://github.com/mozilla/authenticator-rs/pull/337/commits/2a5baf563a61c45720efa1b7783adc58fca501cc

Also applied the same change to other occurrences where applicable: https://github.com/mozilla/authenticator-rs/pull/337/commits/486a7e99f340ccf2cba7c351bed0d34230356958

jschanck commented on src/ctap2/commands/get_assertion.rs:
```diff
@@ -205,6 +320,49 @@ impl GetAssertion {
+                    {
+                        dev.get_shared_secret()
+                            .and_then(|shared_secret| hmac_response.decrypt_secrets(shared_secret))
+                            .and_then(Result::ok)
```
Since we might swallow an error here, it would be nice to add some `debug!()` output to the error paths in `decrypt_secrets`.

emlun replied:
Done! https://github.com/mozilla/authenticator-rs/pull/337/commits/3c04ddabb23310fde06e704175822500a69d5eae

jschanck commented on src/ctap2/mod.rs:
```diff
@@ -672,6 +660,35 @@ pub fn sign<Dev: FidoDevice>(
+                if let Some(secret) = dev.get_shared_secret() {
+                    let (extension, selected_credential) = hmac_get_secret_or_prf.calculate(
+                        secret,
+                        &get_assertion.allow_list,
```
Should we filter down to an allow-list of length 1 before doing this?

emlun replied:
We don't know whether we should before we cross-reference `allow_list` with the extension inputs, so the `calculate` method does that internally.

First, the `sign` function here first filters the `allow_list` to just the credentials applicable to the currently chosen authenticator. At this point the `allow_list` should already be reduced to just 1 element in _most_ cases, as it should be rare for an authenticator to have more than one credential for the same account at the same RP. But then second, `calculate` searches for an `evalByCredential` entry matching any item in `allow_list`. If one is found, `calculate` returns that `PublicKeyCredentialDescriptor` and the result handler here sets `allow_list` to contain only that one item on line 676. Does that seem reasonable?

Perhaps the assignment to `get_assertion.allow_list` is currently a bit hidden. Would it be better to try to move that assignment to the top level?

```diff
diff --git a/src/ctap2/mod.rs b/src/ctap2/mod.rs
index 9894ccf..09996ff 100644
--- a/src/ctap2/mod.rs
+++ b/src/ctap2/mod.rs
@@ -661,28 +661,33 @@ pub fn sign<Dev: FidoDevice>(
         }
 
         // Use the shared secret in the extensions, if requested
-        get_assertion.extensions.hmac_secret = match get_assertion
+        (
+            get_assertion.extensions.hmac_secret,
+            get_assertion.allow_list,
+        ) = match get_assertion
             .extensions
             .hmac_secret
             .take()
             .map(|hmac_get_secret_or_prf| {
                 if let Some(secret) = dev.get_shared_secret() {
-                    let (extension, selected_credential) = hmac_get_secret_or_prf.calculate(
+                    hmac_get_secret_or_prf.calculate(
                         secret,
                         &get_assertion.allow_list,
                         pin_uv_auth_result.get_pin_uv_auth_token(),
-                    )?;
-                    if let Some(selected_credential) = selected_credential {
-                        get_assertion.allow_list = vec![selected_credential.clone()];
-                    }
-                    Ok(extension)
+                    )
                 } else {
-                    Ok(hmac_get_secret_or_prf)
+                    Ok((hmac_get_secret_or_prf, None))
                 }
             })
             .transpose()
         {
-            Ok(extension) => extension,
+            Ok(Some((new_hmac_secret, selected_credential))) => (
+                Some(new_hmac_secret),
+                selected_credential
+                    .map(|selected_credential| vec![selected_credential.clone()])
+                    .unwrap_or(get_assertion.allow_list),
+            ),
+            Ok(None) => (None, get_assertion.allow_list),
             Err(e) => {
                 callback.call(Err(e));
                 return false;
```

jschanck replied:
Ok, that sounds reasonable. I must have overlooked the assignment to `get_assertion.allow_list` on my first pass. No need to move it to the top level.

jschanck reviewed (approved):
This looks great, Emil! My only question is whether we need to do a better job of filtering the allow-list before calling `AuthenticationExtensionsPRFInputs::calculate`. The spec says
> If [evalByCredential](https://w3c.github.io/webauthn/#dom-authenticationextensionsprfinputs-evalbycredential) is present and [contains](https://infra.spec.whatwg.org/#map-exists) an [entry](https://infra.spec.whatwg.org/#map-entry) whose [key](https://infra.spec.whatwg.org/#map-key) is the [base64url encoding](https://w3c.github.io/webauthn/#base64url-encoding) of the [credential ID](https://w3c.github.io/webauthn/#credential-id) that will be returned, let ev be the [value](https://infra.spec.whatwg.org/#map-value) of that entry.

The use of the definite article in "the credential ID that will be returned" makes me think that we should filter down to an allow-list of length 1. Wdyt?

msirringhaus commented on src/crypto/mod.rs:
```diff
@@ -1056,6 +1073,7 @@ impl Serialize for COSEKey {
 
 /// Errors that can be returned from COSE functions.
 #[derive(Debug, Clone, Serialize)]
+#[cfg_attr(test, derive(PartialEq))]
```
I think we can make this struct just PartialEq all the time

emlun replied:
Done! https://github.com/mozilla/authenticator-rs/pull/337/commits/093957b831de339a0b971f91f6aa560353c0ab62

msirringhaus commented on src/ctap2/attestation.rs:
```diff
@@ -1126,4 +1163,140 @@ pub mod test {
+        }
+
+        const PIN_PROTOCOL_2_IV: [u8; 16] = [0; 16]; // PIN protocol 1 uses a hard-coded all-zero IV
+        const EXPECTED_OUTPUT1: [u8; 32] = [
```
If possible, add a comment here how one can verify that this is the expected output (link to spec, if it's in there, or external commands used to calculate it).

emlun replied:
Done! https://github.com/mozilla/authenticator-rs/pull/337/commits/01cb0d351a310ae8edcf7e2c36481b0af979ad4f

msirringhaus commented on src/ctap2/commands/get_assertion.rs:
```diff
@@ -67,17 +68,100 @@ impl UserVerification for GetAssertionOptions {
 }
 
 #[derive(Debug, Clone)]
+#[cfg_attr(test, derive(PartialEq))]
```
Again, `PartialEq` could probably be derived always, unless there is a reason not to. Same below with `HmacGetSecretOrPrf`.

emlun replied:
Done! https://github.com/mozilla/authenticator-rs/pull/337/commits/093957b831de339a0b971f91f6aa560353c0ab62

msirringhaus commented on src/ctap2/server.rs:
```diff
@@ -9,9 +9,12 @@ use serde::{
+use std::collections::HashMap;
 use std::convert::{Into, TryFrom};
 use std::fmt;
 
```
nit: remove blank line here

emlun replied:
Done! https://github.com/mozilla/authenticator-rs/pull/337/commits/5035146fe7a0c6ae4282fbae5c92bca2c0742879

msirringhaus commented on src/ctap2/attestation.rs:
```diff
@@ -1,5 +1,6 @@
+use super::server::HMACGetSecretOutput;
```
I think you can add `HMACGetSecretOutput` 2 lines down, where there are already an server-imports.

emlun replied:
Done! https://github.com/mozilla/authenticator-rs/pull/337/commits/5035146fe7a0c6ae4282fbae5c92bca2c0742879

msirringhaus reviewed (commented):
I haven't looked too deeply at 'the meat' of the PR, but only added a few nitpicks.

emlun commented:
Thanks @jschanck and @msirringhaus for the reviews! I believe I've addressed them all now, and I also noticed a few more things while self-reviewing the new code. While awaiting your re-reviews I'll proceed with testing re-integrating these changes into the Firefox branch.

emlun commented:
I force-pushed a bit because 8bdbc3d16d444ebe250c398d91fa7d1617945844 made the `crypto_dummy` tests fail to compile, and I also took the opportunity to split up 1aa54f08ee738fb7aa99fb11722bf2ab9605eecc into a series of smaller change units: fd8e062a73acdd44098831350b732121c49a1c59, 5cebc5ba24b23886ca3c7aee3bdf77dddf889abb and 9036264bd1b9dd52347197b67fecce2a5b65858a.

In addition to that I also added 978725b534161a9cba107dc0c1bbcadbfc971370 in order to correctly return an empty PRF extension output when the authenticator is not eligible, and added a little more debug output to failure cases.

jschanck commented:
Great! This looks like it's ready to go. Do you want me to merge the `serialize_map!` patch first? I haven't checked whether there are conflicts.

emlun commented:
> Great! This looks like it's ready to go. Do you want me to merge the `serialize_map!` patch first? I haven't checked whether there are conflicts.

Please do! I'll update this PR to make use of it here too.

My local branch also has a few more refactorizations and some additions to the examples. While fiddling with the examples I also found that the (pre-existing) `can_skip_user_verification` logic seems to be a bit strange, but I might be missing something there.

But maybe we'll take all separate from this PR? This one is probably big enough as it is already. :laughing:

jschanck commented:
> Please do! I'll update this PR to make use of it here too.

Sounds good!

> My local branch also has a few more refactorizations and some additions to the examples. While fiddling with the examples I also found that the (pre-existing) `can_skip_user_verification` logic seems to be a bit strange, but I might be missing something there.
> 
> But maybe we'll take all separate from this PR? This one is probably big enough as it is already. 😆

Yes, please submit a separate PR.

Thanks again for your help with this!

emlun commented:
Alright, I've merged the `serialize_map!` changes and added a couple of tests that were now missing. Anything else that needs resolution?

I saw you rebase-merged #338 so at first I tried to rebase this on top of that, but the conflicts were frankly too much work to deal with, and it would have made the history not make sense anymore either, so I decided against it and just merged instead. Is that okay?

irew commented:
Just adding the voice of Dashlane, we would also would like to use PRF over CTAP on Firefox. As of version 136.0.3 on macOS, it seems PRF is not supported.
